### PR TITLE
Fix NLLLoss2d backward int32 offset overflow (#4723)

### LINKED ISSUE #4723: [xpu-aglinment] Fix NLLLoss2d backward 32-bit offset overflow
### 🐛 Describe the bug

`F.nll_loss(...).backward()` produces a silently wrong gradient on XPU when the
flattened offset reaches 2^31. The XPU SYCL backward kernel still uses 32-bit
`sample`, `toffset`, and `ioffset` values, so the offset wraps around instead of
addressing the full tensor.

This was found during XPU alignment for upstream PyTorch issue
https://github.com/pytorch/pytorch/issues/190139. The upstream CUDA path was
already fixed by selecting a 64-bit index type when 32-bit indexing is unsafe;
the analogous XPU kernel was never updated, so XPU still exhibits the original
overflow.

### Reproducer

```python
import torch
import torch.nn.functional as F

print("torch", torch.__version__)
print("xpu available", torch.xpu.is_available())

x = torch.zeros((2**16 + 1, 2**15, 1, 1), device="xpu", dtype=torch.float16, requires_grad=True)
target = torch.zeros((2**16 + 1, 1, 1), device="xpu", dtype=torch.long)

F.nll_loss(x, target, reduction="sum").backward()

assert x.grad[0, 0, 0, 0].item() == -1.0
assert x.grad[-1, 0, 0, 0].item() == -1.0
```

### Actual behavior

The first gradient is correct while the final sample gradient is zero, matching
32-bit offset wraparound at `sample * map_nelem * n_classes == 2**31`:

```text
x.grad[0,0,0,0]  = -1.0
x.grad[-1,0,0,0] = 0.0
RESULT: confirmed
```

### Expected behavior

Every valid target position receives its gradient regardless of tensor size,
matching CPU/CUDA eager semantics:

```text
x.grad[0,0,0,0]  = -1.0
x.grad[-1,0,0,0] = -1.0
```

### Alignment metadata

- Upstream source: https://github.com/pytorch/pytorch/issues/190139
- Source type: upstream issue (CUDA fixed; XPU unaligned)
- Local XPU result: confirmed
- Routed area: `intel/torch-xpu-ops`, XPU NLLLoss2d backward kernel offset typing
- Target repository: `intel-sandbox/torch-xpu-ops-exp`

### Notes

The upstream CUDA issue was closed after its kernel selected a 64-bit index type
when 32-bit indexing is unsafe. The XPU fix should independently align the SYCL
implementation with that guard (widen `sample`/`toffset`/`ioffset` to 64-bit when
the flattened extent exceeds `INT_MAX`) rather than modifying shared PyTorch code.

A regression test should be added under `test/repro/` and run with `pytest`.

### Versions

```text
PyTorch version: 2.14.0.dev20260713+xpu
OS: Ubuntu 22.04.5 LTS (x86_64)
Is XPU available: True
Intel GPU: Intel(R) Data Center GPU Max 1100
Relevant path: src/ATen/native/xpu/sycl/*Loss*.cpp (NLLLoss2d backward)
```

### PR DESCRIPTION
**Pull request by laifenxiawucha**
## Background

XPU NLLLoss2d backward computes per-sample offsets in signed 32-bit integers. Once the flattened offset reaches `2^31`, it wraps and silently leaves gradients for later samples at zero.

## Upstream

- [pytorch/pytorch#190139](https://github.com/pytorch/pytorch/issues/190139) tracks the CUDA overflow.
- [pytorch/pytorch#190144](https://github.com/pytorch/pytorch/pull/190144) and landed [commit 1f18820](https://github.com/pytorch/pytorch/commit/1f18820beba94d55fabe7889653746a5eb7952dd) select 64-bit offsets for the CUDA backward path. XPU has separate offset arithmetic that remained 32-bit.

## Fix

- widen `sample`, `toffset`, and `ioffset` to `int64_t`
- add a regression at the `2^31` offset boundary

## Before and after

| Value | Before | After |
|---|---:|---:|
| First sample gradient | `-1.0` | `-1.0` |
| Last sample gradient | `0.0` | `-1.0` |

## Testing

Intel Data Center GPU Max 1100: the large-offset regression test passed after reproducing the zero tail gradient before the fix.

Fixes #4723

### CODE REVIEW

**Comment by laifenxiawucha** [comment]
## Verification (Intel Data Center GPU Max 1100, source build)

Beyond the exact issue reproducer, ran the checks below to guard against regressions elsewhere.

### 1. XPU vs CPU cross-validation (no CUDA device on this host; CPU is the reference)
Swept `reduction={none,mean,sum}` × `dtype={fp32,fp16,bf16}` × with/without `weight` × with/without `ignore_index`, plus 2**31-scale overflow shapes, comparing forward + backward against the CPU reference element-wise:
- **Gradient (the path this fix touches): 40/40 match CPU.**
- The only 4 loss-value diffs are the fp16 forward **loss** under sum/mean at 2**31-scale extent — pure low-precision accumulation noise in the *unchanged* forward `atomicAdd` path, not the gradient.

### 2. Related PyTorch UTs (via the XPU test harness, `PYTORCH_TESTING_DEVICE_ONLY_FOR=xpu`)
- `test_nn_xpu.py -k "nll_loss and xpu"` → **17 passed**
- `test_nn_xpu.py -k "(NLLLoss|nll|cross_entropy|CrossEntropy) ..."` → **57 passed, 33 skipped**
- `test_ops_gradients_xpu.py -k nll_loss` → **6 passed, 9 skipped**

### 3. Regression test
Added `test_nll_loss2d_large_tensor` to `test/xpu/test_nn_xpu.py`, mirroring the existing CUDA-style `test_nll_loss_large_tensor`. The existing test covers the **1D** `nll_loss` path (2D input); this one exercises the **`nll_loss2d`** path (4D input) that this fix touches, which previously had no large-tensor coverage.
`fp32`, `>2**31` elements, compared to CPU for all three reductions — **3/3 passed** post-fix.

File: test/repro/test_nll_loss2d_backward_offset_overflow.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+import unittest
+
+import torch
+import torch.nn.functional as F
+from torch.testing._internal.common_utils import run_tests, TestCase
+
+
+@unittest.skipIf(not torch.xpu.is_available(), "XPU not available")
+class TestNllLoss2dBackwardOffsetOverflow(TestCase):
```
[the remark this answers is not in the document]
Comment (laifenxiawucha) [reply test/repro/test_nll_loss2d_backward_offset_overflow.py:29]: done

**Review by ZhaoqiongZ — CHANGES_REQUESTED** [review]
please fix the CI failure

**Comment by ZhaoqiongZ** [comment]
@copilot+claude-opus-4.5 fix the lint CI failure
